doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,396 @@ Applying theorems
+
+      :tacn:`apply` unifies the conclusion `n <= p` of the theorem
+      `le_trans : forall n m p, n <= m -> m <= p -> n <= p`
+      with the goal, assigning `x * x` and `y * y` in the goal
+      to, repectively, `n` and `p` in theorem (backward reasoning).
+      The `with` clause provides the binding for `m`:
+
+      .. coqtop:: reset in
+
+         Require Import PeanoNat.
+
+      .. coqtop:: none
 
-   .. tacv:: apply @term with @bindings
+         Goal forall (x y : nat), x <= y -> x * x <= y * y.
 
-      This also provides apply with values for instantiating premises. Here, variables
-      are referred by names and non-dependent products by increasing numbers (see
-      :ref:`bindings`).
+      .. coqtop:: out
+
+         intros x y H0.
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (y * x).
+
+   .. _apply_with_binding_hyp:
+   .. example:: Apply a theorem with a binding in a hypothesis
+
+      When applying a theorem in a hypothesis,
+      :tacn:`apply` unifies the hypothesis with one of the premises
+      of the theorem `le_trans : forall n m p, n <= m -> m <= p -> n <= p`.
+      In this case, it unifies with the first premise
+      (`n <= m`) and assigns `x * x` and `y * y` to,
+      respectively, `n` and `m` in the theorem (forward reasoning).
+      The  `with` clause provides the binding for `p`.
+
+      In addition,
+      `apply` in a hypothesis isn't as flexible as
+      `apply` in the goal: for hypotheses, the unbound variable can be bound
+      by name (as shown) or values for all the variables can be given
+      positionally,
+      i.e. `apply Nat.le_trans with (x * x) (y * y) (y * x) in H.`
+
+      .. coqtop:: reset in
+
+         Require Import PeanoNat.
+
+      .. coqtop:: none
+
+         Goal forall (x y : nat), x * x <= y * y -> x <= y.
+
+      .. coqtop:: out
+
+         intros x y H.
+
+      .. coqtop:: all
+
+         apply Nat.le_trans with (p := y * x) in H.
+
+   .. _apply_with_iff:
+   .. example:: Applying theorems with `<->`
+
+      Theorems that use `<->` to state a logical equivalence behave
+      consistently when applied to goals and hypotheses; the tactic
+      picks the useful alternative.
+
+      .. coqtop:: reset none
+
+         Lemma iff: 1 = 1 <-> 2 = 2.  Admitted.
+         Goal forall (H: 1 = 1), 1 = 1.
+
+      .. coqtop:: out
+
+         intros H.
+
+      .. coqtop:: all
+
+         apply iff.
+         apply iff in H.
+
+   .. _example_apply_hyp_inductive:
+   .. example:: apply an inductive type in a hypothesis
 
-      .. flag:: Apply With Renaming
+      .. coqtop:: reset none
+
+         Goal forall A B C: Prop, (A <-> B) -> A -> False.
+
+      .. coqtop:: out
+
+         intros A B C H1 H2.
+
+      .. coqtop:: all
+
+         apply H1 in H2.
+
+   .. _example_apply_pattern:
+   .. example:: Using :tacn:`pattern` to control unification in apply
+
+      After the ``pattern``, the ``apply`` uses second-order unification.
```

Comment:
Depending on how exactly we define second-order unification, we could also consider that this is already second-order unification before the call to pattern (since we unify to what is, semantically, a function). Or we could consider that even after the pattern, this is not second-order unification (maybe the first-order unification algorithm as implemented suffices in this case and the second-order unification mechanism is only used when it needs to create a `fun`). If we want to know, we need to ask unification experts such as @herbelin.

Reply:
How's this?  It's definitely a second order match but I think this is clear that it's not a general purpose second-order algorithm.  IIUC it's undecidable.

      After the ``pattern``, the ``apply`` uses the special case second-order
      unification described :ref:`here <apply_with_second_order_unification>`
      (after "unless").